### PR TITLE
Incentive curve

### PR DESCRIPTION
**Pull request by dharkmattr**

### CODE REVIEW

File: contracts/masset/Masset.sol [hunk 1]
```diff
@@ -233,6 +308,8 @@ contract Masset is IERC777Recipient, InitializableOwnable, InitializableReentran
             IERC20(_basset).transfer(_recipient, bassetQuantity);
         }
 
+        mintPenalty(_basset, bassetQuantity, _recipient);
```
Comment (jameshowlett977) [inline contracts/masset/Masset.sol:311]: I haven't found a transfer of XUSD from the user's account as supposed according to spec, but we mint tokens to RewardManager.

File: contracts/masset/RewardManager.sol [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    /// @return massetAddress the masset address
+    function getMassetAddress() external view returns (address) {
+        return address(masset);
+    }
+
+    /** Setters **/
+
+    /// @notice Set the factor
+    /// @param _factor the factor
+    function setFactor(uint256 _factor) external onlyOwner {
```
Comment (jameshowlett977) [inline contracts/masset/RewardManager.sol:81]: _BabelFish Bitocracy should be able to add and remove addresses from an "incentive administrator" list. Every address that is on this list is an "incentive administrator"._

I think we need an admin role, admins should be able to invoke "setFactor" and "setTargetWeights".

File: contracts/masset/RewardManager.sol [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        address _recipient, uint256 _minimum) external returns (uint256) {
+
+        require(msg.sender == address(masset), "only masset may call");
+
+        IBasketManager basketManager = IBasketManager(masset.getBasketManager());
+        IERC20 XUSD = IERC20(masset.getToken());
+
+        uint256 reward = this.getRewardForDeposit(_bassetAddress, _sum);
+        uint256 available = XUSD.balanceOf(address(this));
+        reward = reward < available ? reward : available;
```
Comment (jameshowlett977) [inline contracts/masset/RewardManager.sol:196]: Does it mean no rewards for the user even if he deserves it just because we don’t have enough balance on RewardManager?
Shouldn't we save rewards for users in order to withdraw later when funds are added?

File: contracts/masset/RewardManager.sol [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+        require(reward >= _minimum, "reward under minimum");
+
+        XUSD.transfer(_recipient, reward);
```
Comment (jameshowlett977) [inline contracts/masset/RewardManager.sol:200]: Please add "require("

File: test/RewardManager.spec.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+import { ZERO_ADDRESS } from "@utils/constants";
+
+const { expect } = envSetup.configure();
+
+const RewardManager = artifacts.require("RewardManager");
+const Masset = artifacts.require("Masset");
+const MockERC20 = artifacts.require("MockERC20");
+const MockBasketManager = artifacts.require("MockBasketManager");
+
+const testCases = [
```
Comment (jameshowlett977) [inline test/RewardManager.spec.ts:16]: Please add unit tests for all functions and negative cases for already added.

File: contracts/helpers/MappingAddressToUint256.sol [hunk 6]
```diff
@@ -0,0 +1,63 @@
+pragma solidity 0.5.17;
+
+import "../helpers/Ownable.sol";
+
+contract MappingAddressToUint256 is Ownable {
```
Comment (jameshowlett977) [inline contracts/helpers/MappingAddressToUint256.sol:5]: Please add tests for this contract.

**Review by jameshowlett977 — CHANGES_REQUESTED** [review]
Please add tests for MappingAddressToUint256, all other comments related to this contract are minor, we can leave as is, but should be aware of.

File: contracts/helpers/MappingAddressToUint256.sol [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    function getKeys() public view returns(address[] memory) {
+        return keys;
+    }
+
+    function getValues() public view returns(uint256[] memory) {
+        uint256[] memory r = new uint256[](keys.length);
+        for(uint i=0; i<r.length; i++) {
+            r[i] = values[keys[i]];
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/helpers/MappingAddressToUint256.sol:27]: I don't think so.... if the key is removed then the value will not be returned.

File: contracts/helpers/MappingAddressToUint256.sol [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    function remove(address _address) public onlyOwner {
+        if(!flags[_address]) {
+            return;
+        }
+        flags[_address] = false;
+        values[_address] = 0;
+        bool f = false;
+        for(uint i=0; i < keys.length; i++) {
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/helpers/MappingAddressToUint256.sol:53]: No, order is not important.... fixed.

File: contracts/masset/BasketManager.sol [hunk 9]
```diff
@@ -66,11 +80,61 @@ contract BasketManager {
...
+        emit onBassetAdded(msg.sender, _basset, _factor, _bridge);
+    }
+
+    function removeBasset(address _basset) public onlyOwner {
+        require(_isValidBasset(_basset), "invalid _basset");
+        bassetsMap[_basset] = false;
+        emit onBassetRemoved(msg.sender, _basset);
+        for(uint i=0; i<bassetsArray.length; i++) {
+            if(bassetsArray[i] == _basset) {
+                bassetsArray[i] = bassetsArray[bassetsArray.length-1];
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/masset/BasketManager.sol:134]: is a redundant a[i] = a[i] more expensive than testing for i!=a.length-1 ?
I'm not sure....
At any rate this function is not likely to be called often, so gas cost is not important.

File: contracts/masset/Masset.sol (hunk 1 above)
Comment (dharkmattr) [reply contracts/masset/Masset.sol:311]: instead of transferring from the user's account, we burn all the XUSD and then mist the reward directly to the reward manager. I think this is the same thing, and it makes the logic much more simple.

File: contracts/masset/RewardManager.sol (hunk 2 above)
Comment (dharkmattr) [reply contracts/masset/RewardManager.sol:81]: mmm... why do we not use the same multisig for this as all the other parameters in the system? I don't trust any one person to set the factor or the weights, in case there's a large amount of funds in the reward manager.

File: contracts/masset/RewardManager.sol (hunk 3 above)
Comment (dharkmattr) [reply contracts/masset/RewardManager.sol:196]: The design was that if the funds run out we just don't incentivize. The UI will reflect this so nobody will expect a reward and not get it. If we save negative amounts and allow them to be withdrawn later it will cause a run on the funds when they come in, which I think is unfair and ungainly. Also, no such UI has been planned (to withdraw rewards at a later time).

File: contracts/helpers/MappingAddressToUint256.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (jameshowlett977) [reply contracts/helpers/MappingAddressToUint256.sol:27]: You're right.

File: contracts/masset/Masset.sol (hunk 1 above)
Comment (jameshowlett977) [reply contracts/masset/Masset.sol:311]: agree

File: contracts/masset/RewardManager.sol (hunk 2 above)
Comment (jameshowlett977) [reply contracts/masset/RewardManager.sol:81]: The spec should be updated then.

File: contracts/masset/RewardManager.sol (hunk 3 above)
Comment (jameshowlett977) [reply contracts/masset/RewardManager.sol:196]: Got it, just make sure UI clearly tells about available reward.

File: contracts/masset/RewardManager.sol (hunk 2 above)
Comment (dharkmattr) [reply contracts/masset/RewardManager.sol:81]: It just occurred to me if we want this feature we can just set up a multisig with a 1 signer threshold and set bitocracy as its owner. So this capability is not in the scope of this PR.

Comment (dharkmattr) [reply contracts/masset/RewardManager.sol:81]: light agrees to change the spec

File: contracts/helpers/MappingAddressToUint256.sol [hunk 10]
```diff
@@ -0,0 +1,60 @@
+pragma solidity 0.5.17;
+
+import { Ownable } from "../openzeppelin/contracts/ownership/Ownable.sol";
+
+contract MappingAddressToUint256 is Ownable {
```
Comment (ororopickpocket) [inline contracts/helpers/MappingAddressToUint256.sol:5]: is there a reason why you're not using open zeppelins enumerable map for this? it does support address => uint 
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableMap.sol

File: contracts/helpers/MappingAddressToUint256.sol [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    mapping(address => uint256) private values;
+    mapping(address => bool) private flags;
```
Comment (ororopickpocket) [inline contracts/helpers/MappingAddressToUint256.sol:8]: Is it important to have the flag? 
As i understand it, you're using the flag only to know which key has a value set. 
Why not just check if values[address] == 0 ?
Do you need to know if an address is consciously mapped to 0?

File: contracts/helpers/MappingAddressToUint256.sol [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        return values[_address];
+    }
+
+    function getKeys() public view returns(address[] memory) {
+        return keys;
+    }
+
+    function getValues() public view returns(uint256[] memory) {
+        uint256[] memory r = new uint256[](keys.length);
+        for(uint i=0; i<r.length; i++) {
```
Comment (ororopickpocket) [inline contracts/helpers/MappingAddressToUint256.sol:26]: depending on the number of keys this could become a very long loop up to the point where it cannot be processed in a single transaction. there should be a safeguard either by limiting the length of `keys` or by limiting the number of iterations here or by allowing the caller to pass start and end indexes. 
should you decide to do none of these because it seems not important to you in this case, at least add a comment in the function documentation informing developers that this needs to be considered.

File: contracts/helpers/MappingAddressToUint256.sol [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        }
+    }
+
+    function remove(address _address) public onlyOwner {
+        if(!flags[_address]) {
+            return;
+        }
+        flags[_address] = false;
+        values[_address] = 0;
+        for(uint i=0; i < keys.length; i++) {
```
Comment (ororopickpocket) [inline contracts/helpers/MappingAddressToUint256.sol:52]: again, this can become very expensive or impossible to execute if `keys` becomes too large. 
removing could be cheaper if you would keep an index. the index could be a mapping (address => uint) which lets you know at which point in the array, the key is stored. if you start counting the indexes from 1, you can replace the flag mapping with it, because then you can check if a key has been set by looking up its index (index 0 would mean not set).

**Review by ororopickpocket — COMMENTED** [review]
So far I only reviewed MappingAddressToUint256. I added a few comments for discussion.
I will continue the review later and/or tomorrow.

**Review by ororopickpocket — COMMENTED** [review]
This is the review of Masset.sol only. The review of the RewardsManager is still outstanding.

File: contracts/helpers/MappingAddressToUint256.sol (hunk 11 above)
Comment (dharkmattr) [reply contracts/helpers/MappingAddressToUint256.sol:8]: yes, a value = 0 is not the same as a token doesn't exist. I could have used a +1 trick, so that 0 = n/a and 1 = 0, but I think making the code less readable is not worth the gas gained from such optimization for small number of values

File: contracts/helpers/MappingAddressToUint256.sol (hunk 13 above)
Comment (dharkmattr) [reply contracts/helpers/MappingAddressToUint256.sol:52]: removing a token is very rare. Actually, I don't even think it will ever happen

File: contracts/masset/BasketManager.sol [hunk 14]
```diff
@@ -66,11 +80,61 @@ contract BasketManager {
...
         return bridgeMap[_basset];
     }
 
-    function getVersion() external pure returns(string memory) {
-        return "2.0";
+    function getVersion() public pure returns(string memory) {
+        return "3.0";
+    }
+
+    function getBalanceInMasset(address _massetAddress, address _bassetAddress) public view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/masset/BasketManager.sol:91]: Well, every token can in principal have a different precision. This function gives you the balance of a basset in masset units. The code can be quite complex at times because every calculation has to convert amounts from basset units to masset units and vice versa.

In retrospect, all of the tokens we've ever had, had 18 digit precision, so this code can be simplified by removing all of this... this is a topic for future discussion.

File: contracts/helpers/MappingAddressToUint256.sol (hunk 10 above)
Comment (dharkmattr) [reply contracts/helpers/MappingAddressToUint256.sol:5]: I have philosophical objections to their implementation. I think it's unduly complicated, and I strongly believe complicated code === insecure code. I'm suspicious of oenzeppelin because of their horrible horrible upgradable proxy pattern. In the beginning I chose to use it because I thought "it's from openzeppelin so it must be good", but I'm regretting that choice almost daily now.

File: contracts/masset/RewardManager.sol [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        uint256 _sum
+        ) external view returns (uint256) {
+
+        // uninitialized! avoid division by zero
+        require(targetWeights.getKeys().length > 0, "no target weights");
+
+        address[] memory tokens = targetWeights.getKeys();
+        (uint256 dsqrBefore, uint256 dsqrAfter) = getAverageDsqrs(_bassetAddress, 0 - int256(_sum));
+        uint256 penalty = dsqrAfter > dsqrBefore ?
+            factor.mul(dsqrAfter - dsqrBefore).div(tokens.length).div(ONE) :
```
Comment (ororopickpocket) [inline contracts/masset/RewardManager.sol:178]: this appears to be even more incorrect, because you're dividing by tokens.length which you already did in getAverageDsqrs

**Review by ororopickpocket — CHANGES_REQUESTED** [review]
Please check the math and add actual test cases for the reward and penalty calculations.
Also, natspec documentation would be nice.

File: contracts/masset/RewardManager.sol (hunk 15 above)
Comment (dharkmattr) [reply contracts/masset/RewardManager.sol:178]: Since factor is an arbitrary number to be determined by trial and error, it doesn't really matter. The question is if factor is defined in total or per stable-coin.

File: contracts/masset/BasketManager.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/masset/BasketManager.sol:91]: @ororopickpocket your thoughts?

File: contracts/masset/RewardManager.sol [hunk 16]
```diff
@@ -142,6 +149,9 @@ contract RewardManager is IRewardManager, Ownable {
         dsqrAfter = dsqrAfter / tokens.length;
     }
 
+    /// @notice Get the reward for a deposit
+    /// @param _bassetAddress the token addresses
+    /// @param _sum the amount to deposit
     function getRewardForDeposit(
```
Comment (ororopickpocket) [inline contracts/masset/RewardManager.sol:155]: why make this one external if you're calling it internally as well?
why not public ?

File: contracts/masset/RewardManager.sol [hunk 17]
```diff
@@ -154,16 +164,25 @@ contract RewardManager is IRewardManager, Ownable {
...
+
+        // the asymptotic part
+        uint256 vBefore = dsqrBefore.mul(ONE).div(ONE.sub(dsqrBefore));
+        uint256 vAfter = dsqrAfter.mul(ONE).div(ONE.sub(dsqrAfter));
+
+        uint256 reward = vAfter < vBefore ?
+            factor.mul(vBefore.sub(vAfter)).div(ONE) :
             0;
         reward = reward.mul(_sum).div(ONE);
+        reward = reward < _sum ? reward : _sum;
```
Comment (ororopickpocket) [inline contracts/masset/RewardManager.sol:176]: what's the purpose of this line?
I see that it's capping the maximum reward to the basset amount being deposited, but i don't see this on the spec.

File: contracts/masset/Masset.sol [hunk 18]
```diff
@@ -108,10 +106,13 @@ contract Masset is IMasset, IERC777Recipient, InitializableOwnable, Initializabl
         version = "1.0";
     }
 
-    /** Incentive functions */
+    /**
+    * Internal incentive functions
+    **/
```
Comment (ororopickpocket) [inline contracts/masset/Masset.sol:111]: payReward is called by `onTokensMinted` and `_mintTo`. The first calls it after the basset is already transferred to the contract, the latter before it is transferred. This affects the outcome of `getAverageDsqrs`. Hadn't you planned to change this? Am I missing something?

Comment (dharkmattr) [reply contracts/masset/Masset.sol:111]: I put this correction in a separate PR and sent it to you. I only now realized you must have missed it, sorry for not communicating more clearly.

File: contracts/masset/Masset.sol [hunk 19]
```diff
@@ -369,12 +515,18 @@ contract Masset is IERC777Recipient, InitializableOwnable, InitializableReentran
...
+        require(_newRewardManager == address(0) || Address.isContract(_newRewardManager), "not a contract");
+        emit onSetRewardManager(msg.sender, address(rewardManager), _newRewardManager);
+        rewardManager = IRewardManager(_newRewardManager);
+    }
+
     // Temporary migration code
 
-    function migrateToV3(address _pauseManager) public onlyOwner {
-        require(keccak256(bytes(version)) == keccak256(bytes("2.10")), "wrong version");
+    function migrateToV4() public {
```
Comment (koirikivi) [inline contracts/masset/Masset.sol:527]: Just noticed this, but is it intentional that anyone can migrate?

Comment (dharkmattr) [reply contracts/masset/Masset.sol:527]: yes, this can only happens once and we don't care who calls it as long as it gets called.

File: contracts/masset/RewardManager.sol (hunk 17 above)
Comment (dharkmattr) [reply contracts/masset/RewardManager.sol:240]: It's here just in case. In practice there's a maximum possible reward parameter that will never be over 100%. If this is not in the spec we should update the spec.
